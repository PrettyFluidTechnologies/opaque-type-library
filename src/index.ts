export type Opaque<K extends string> = {
  readonly __OPAQUE__: '__OPAQUE__';
  readonly __OPAQUE_KEY__: K;
  readonly value: unknown;
};

export function createOpaqueAPI<Key extends string, $PrivateValue>(key: Key) {
  type LocalOpaque = Opaque<Key>;

  return {
    toOpaque(value: $PrivateValue): LocalOpaque {
      return {
        __OPAQUE__: '__OPAQUE__',
        __OPAQUE_KEY__: key,
        value,
      };
    },

    fromOpaque(opaque: LocalOpaque): $PrivateValue {
      // we force the return type here, we're sure this is $PrivateValue
      return opaque.value as $PrivateValue;
    },

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    isOpaque(thing: any): thing is LocalOpaque {
      // eslint-disable-next-line no-underscore-dangle
      return thing.__OPAQUE__ === '__OPAQUE__' && thing.__OPAQUE_KEY__ === key;
    },
  }
}
