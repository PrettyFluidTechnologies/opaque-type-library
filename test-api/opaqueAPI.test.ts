import { createOpaqueAPI } from '../src/index';

type $Private = number;

describe('createOpaqueAPI', () => {
  const { toOpaque, fromOpaque, isOpaque } = createOpaqueAPI<'Thing', $Private>(
    'Thing',
  );

  describe('toOpaque', () => {
    it('should return an opaque', () => {
      const opaqueThing = toOpaque(2);

      expect(opaqueThing).toMatchSnapshot();
    });
  });

  describe('fromOpaque', () => {
    it('should return the private value', () => {
      const opaqueThing = toOpaque(2);
      const valueThing = fromOpaque(opaqueThing);

      expect(valueThing).toEqual(2);
    });
  });

  describe('isOpaque', () => {
    it('should return the private value', () => {
      expect(isOpaque(toOpaque(2))).toEqual(true);
      expect(isOpaque('dfhsajdlf')).toEqual(false);
    });
  });
});
