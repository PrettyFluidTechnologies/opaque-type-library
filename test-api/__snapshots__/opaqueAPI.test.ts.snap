// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createOpaqueAPI toOpaque should return an opaque 1`] = `
Object {
  "__OPAQUE_KEY__": "Thing",
  "__OPAQUE__": "__OPAQUE__",
  "value": 2,
}
`;
